File: system_monitor.js
// system_monitor.js
const os = require('os');
const { exec } = require('child_process');
const readline = require('readline');

readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);

class SystemMonitor {
    constructor(interval = 2) {
        this.interval = interval;
        this.running = true;
        this.logging = false;
        this.showProcesses = false;
    }

    getCPUUsage() {
        const cpus = os.cpus();
        let idle = 0, total = 0;
        cpus.forEach(cpu => {
            for (let type in cpu.times) {
                total += cpu.times[type];
            }
            idle += cpu.times.idle;
        });
        return { idle, total };
    }

    getMemoryUsage() {
        const total = os.totalmem();
        const free = os.freemem();
        const used = total - free;
        const percent = (used / total) * 100;
        return { total, used, free, percent };
    }

    getTopProcesses(callback) {
        const isWin = process.platform === 'win32';
        const cmd = isWin ? 'tasklist /fo csv /nh' : 'ps aux --sort=-%cpu | head -6';
        exec(cmd, (err, stdout) => {
            if (err) {
                callback([]);
                return;
            }
            const procs = [];
            const lines = stdout.split('\n');
            for (let line of lines) {
                if (!line.trim()) continue;
                if (isWin) {
                    const parts = line.split(',').map(s => s.trim().replace(/"/g, ''));
                    if (parts.length >= 5) {
                        const name = parts[0];
                        const pid = parseInt(parts[1]);
                        const mem = parseFloat(parts[4].replace(/,/g, '').replace('K', '')) / 1024;
                        procs.push({ pid, name, cpu: 0, mem });
                    }
                } else {
                    const fields = line.trim().split(/\s+/);
                    if (fields.length >= 11 && fields[0] !== 'USER') {
                        const pid = parseInt(fields[1]);
                        const cpu = parseFloat(fields[2]);
                        const mem = parseFloat(fields[3]);
                        const name = fields.slice(10).join(' ');
                        procs.push({ pid, name, cpu, mem });
                    }
                }
            }
            callback(procs.slice(0, 5));
        });
    }

    formatBar(percent, width = 30) {
        const filled = Math.round((percent / 100) * width);
        return '█'.repeat(filled) + '░'.repeat(width - filled);
    }

    render() {
        const cpu = this.getCPUUsage();
        const mem = this.getMemoryUsage();
        const memTotalGB = mem.total / 1024**3;
        const memUsedGB = mem.used / 1024**3;

        console.clear();
        console.log('=== System Monitor ===');
        console.log(`Refresh interval: ${this.interval}s (press 'q' to quit, '+'/'-' to adjust, 'l' log, 'p' processes)`);
        console.log();
        const cpuPercent = 100 - (cpu.idle / cpu.total * 100);
        const cpuBar = this.formatBar(cpuPercent);
        const memBar = this.formatBar(mem.percent);
        console.log(`CPU:   ${cpuBar}  ${cpuPercent.toFixed(1)}%`);
        console.log(`RAM:   ${memBar}  ${mem.percent.toFixed(1)}%  (Used: ${memUsedGB.toFixed(1)} GB / Total: ${memTotalGB.toFixed(1)} GB)`);

        if (this.showProcesses) {
            console.log('\nTop processes by CPU:');
            this.getTopProcesses((procs) => {
                if (procs.length === 0) {
                    console.log('(No data)');
                } else {
                    console.log('PID     NAME                 CPU%    MEM%');
                    procs.forEach(p => {
                        console.log(`${String(p.pid).padEnd(8)} ${p.name.slice(0, 20).padEnd(20)} ${String(p.cpu ? p.cpu.toFixed(1) : '0.0').padEnd(8)} ${String(p.mem ? p.mem.toFixed(1) : '0.0').padEnd(8)}`);
                    });
                }
            });
        }
    }

    run() {
        console.log('Starting System Monitor... (press q to quit)');
        process.stdin.on('keypress', (ch, key) => {
            if (key && key.name === 'q') {
                this.running = false;
                process.stdin.setRawMode(false);
                process.stdin.pause();
                console.log('\nGoodbye!');
                process.exit(0);
            }
            if (key && key.name === '+') {
                this.interval = Math.min(this.interval + 1, 10);
            }
            if (key && key.name === '-') {
                this.interval = Math.max(this.interval - 1, 1);
            }
            if (key && key.name === 'l') {
                this.logging = !this.logging;
                console.log('\nLogging toggled.');
            }
            if (key && key.name === 'p') {
                this.showProcesses = !this.showProcesses;
            }
        });

        const loop = () => {
            if (!this.running) return;
            this.render();
            setTimeout(loop, this.interval * 1000);
        };
        loop();
    }
}

const monitor = new SystemMonitor(2);
monitor.run();

File: system_monitor.py
# system_monitor.py
import os
import sys
import time
import threading
import platform
import csv
from datetime import datetime

try:
    import psutil
except ImportError:
    print("Please install psutil: pip install psutil")
    sys.exit(1)

class SystemMonitor:
    def __init__(self, interval=2):
        self.interval = interval
        self.running = True
        self.logging = False
        self.log_file = None
        self.show_processes = False
        self.lock = threading.Lock()

    def get_cpu_usage(self):
        return psutil.cpu_percent(interval=0.1, percpu=False)

    def get_cpu_per_core(self):
        return psutil.cpu_percent(interval=0.1, percpu=True)

    def get_memory_usage(self):
        mem = psutil.virtual_memory()
        return {
            'total': mem.total,
            'used': mem.used,
            'free': mem.free,
            'percent': mem.percent
        }

    def get_top_processes(self, limit=5):
        procs = []
        for p in psutil.process_iter(['pid', 'name', 'cpu_percent', 'memory_percent']):
            try:
                info = p.info
                procs.append(info)
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                pass
        procs.sort(key=lambda x: x['cpu_percent'] or 0, reverse=True)
        return procs[:limit]

    def format_bar(self, percent, width=30):
        filled = int(width * percent / 100)
        bar = '█' * filled + '░' * (width - filled)
        return bar

    def render(self):
        cpu = self.get_cpu_usage()
        mem = self.get_memory_usage()
        mem_percent = mem['percent']
        mem_used_gb = mem['used'] / (1024**3)
        mem_total_gb = mem['total'] / (1024**3)

        os.system('cls' if platform.system() == 'Windows' else 'clear')
        print("=== System Monitor ===")
        print(f"Refresh interval: {self.interval}s (press 'q' to quit, '+'/'-' to adjust, 'l' to toggle log, 'p' to toggle processes)")
        print()
        cpu_bar = self.format_bar(cpu)
        mem_bar = self.format_bar(mem_percent)
        print(f"CPU:   {cpu_bar}  {cpu:5.1f}%")
        print(f"RAM:   {mem_bar}  {mem_percent:5.1f}%  (Used: {mem_used_gb:.1f} GB / Total: {mem_total_gb:.1f} GB)")

        if self.show_processes:
            print("\nTop processes by CPU:")
            procs = self.get_top_processes(5)
            print(f"{'PID':<8} {'NAME':<20} {'CPU%':<8} {'MEM%':<8}")
            for p in procs:
                print(f"{p['pid']:<8} {p['name'][:20]:<20} {p['cpu_percent'] or 0:<8.1f} {p['memory_percent'] or 0:<8.1f}")

        if self.logging:
            with self.lock:
                if self.log_file is None:
                    self.log_file = open('system_monitor.csv', 'w', newline='')
                    writer = csv.writer(self.log_file)
                    writer.writerow(['timestamp', 'cpu_percent', 'mem_percent', 'mem_used_gb', 'mem_total_gb'])
                else:
                    writer = csv.writer(self.log_file)
                writer.writerow([
                    datetime.now().isoformat(),
                    round(cpu, 1),
                    round(mem_percent, 1),
                    round(mem_used_gb, 2),
                    round(mem_total_gb, 2)
                ])
                self.log_file.flush()

    def run(self):
        print("Starting System Monitor... (press 'q' to quit)")
        # Use a separate thread to listen for keypresses
        def input_listener():
            while self.running:
                try:
                    ch = sys.stdin.read(1)
                    if ch == 'q':
                        self.running = False
                    elif ch == '+':
                        self.interval = min(self.interval + 1, 10)
                    elif ch == '-':
                        self.interval = max(self.interval - 1, 1)
                    elif ch == 'l':
                        self.logging = not self.logging
                        if self.logging:
                            print("\nLogging started.")
                        else:
                            print("\nLogging stopped.")
                            with self.lock:
                                if self.log_file:
                                    self.log_file.close()
                                    self.log_file = None
                    elif ch == 'p':
                        self.show_processes = not self.show_processes
                except:
                    pass

        thread = threading.Thread(target=input_listener, daemon=True)
        thread.start()

        while self.running:
            self.render()
            time.sleep(self.interval)

        if self.log_file:
            self.log_file.close()
        print("Goodbye!")

if __name__ == '__main__':
    monitor = SystemMonitor()
    monitor.run()
